### DISCUSSION TITLE
Announcements and New Features (2021)

### DISCUSSION THREAD

**Discussion by epezent** [epezent/implot#168] [Announcements]
Announcements of changes and new features will be posted here. Click `Subscribe` to the right to receive updates!

**Feel free to comment on announcements, but please do not clutter the main thread with comments.**

Previous Announcements:
[Announcements and New Features (2020)](https://github.com/epezent/implot/issues/48)

**Comment by epezent** [comment]
## Labels for Auxiliary Y-Axes - (v0.9 - 1/6/2021)

Labels can now be added to the second and third y-axes by passing strings to `BeginPlot`. Additionally, `ImPlotAxisFlags_NoLabel` has been added and a context menu toggle is provided. Special thanks to @mindv0rtex!

![](https://user-images.githubusercontent.com/10409817/103726193-525e2880-4fa6-11eb-843a-7834bbda4880.png)

**Comment by epezent** [comment]
## PlotVLines / PlotHLines - (v0.9 - 1/15/2021)

Horizontal and vertical reference lines that extend to +/- infinity can now be added with the following functions:

```cpp
void PlotVLines(const char* label_id, const T* xs, int count, int offset=0, int stride=sizeof(T));
void PlotHLines(const char* label_id, const T* ys, int count, int offset=0, int stride=sizeof(T));
```

![inf](https://user-images.githubusercontent.com/29577475/104697638-af0fc080-56d5-11eb-9f2a-29673fe59eec.gif)

**Comment by epezent** [comment]
## Single Right-Click Context Menus - (v0.9 - 1/18/2021)

Plot context menus now open with _single_ right-click. This is obviously more natural than the previous double right click requirement. See #170.

Additionally, `ImPlotInputMap` has been moved out of the public API and into `implot_internal.h`. As far as I know, the feature is not widely used, and currently input handling is not mature enough to catch all of the edge cases that may occur under certain configurations of the input mapping. As such, this feature should be considered experimental/WIP and used with caution. If you were previously using it, simply include `implot_internal.h` to continue as you were.

**Comment by epezent** [comment]
## Improve Drag and Drop Support - (v0.9 - 2/28/2021)

Drag and drop support has been improved. Now, plot areas, axes, legends, and items can serve as either a target OR source. Get creative! 

```cpp
// Turns the current plot's plotting area into a drag and drop target. Don't forget to call EndDragDropTarget!
bool BeginDragDropTarget();
// Turns the current plot's X-axis into a drag and drop target. Don't forget to call EndDragDropTarget!
bool BeginDragDropTargetX();
// Turns the current plot's Y-Axis into a drag and drop target. Don't forget to call EndDragDropTarget!
bool BeginDragDropTargetY(ImPlotYAxis axis = ImPlotYAxis_1);
// Turns the current plot's legend into a drag and drop target. Don't forget to call EndDragDropTarget!
bool BeginDragDropTargetLegend();
// Ends a drag and drop target (currently just an alias for ImGui::EndDragDropTarget).
void EndDragDropTarget();

// NB: By default, plot and axes drag and drop sources require holding the Ctrl modifier to initiate the drag.
// You can change the modifier if desired. If ImGuiKeyModFlags_None is provided, the axes will be locked from panning.

// Turns the current plot's plotting area into a drag and drop source. Don't forget to call EndDragDropSource!
bool BeginDragDropSource(ImGuiKeyModFlags key_mods = ImGuiKeyModFlags_Ctrl, ImGuiDragDropFlags flags = 0);
// Turns the current plot's X-axis into a drag and drop source. Don't forget to call EndDragDropSource!
bool BeginDragDropSourceX(ImGuiKeyModFlags key_mods = ImGuiKeyModFlags_Ctrl, ImGuiDragDropFlags flags = 0);
// Turns the current plot's Y-axis into a drag and drop source. Don't forget to call EndDragDropSource!
bool BeginDragDropSourceY(ImPlotYAxis axis = ImPlotYAxis_1, ImGuiKeyModFlags key_mods = ImGuiKeyModFlags_Ctrl, ImGuiDragDropFlags flags = 0);
// Turns an item in the current plot's legend into drag and drop source. Don't forget to call EndDragDropSource!
bool BeginDragDropSourceItem(const char* label_id, ImGuiDragDropFlags flags = 0);
// Ends a drag and drop source (currently just an alias for ImGui::EndDragDropSource).
void EndDragDropSource();
```

![](https://user-images.githubusercontent.com/29577475/105454558-b25ffa80-5c47-11eb-8c44-406a90ae7a1f.gif)

**Reply by djbarker** [reply]
Hi! I know this post is a few years old, but wondering if you could share the code for this drag-and-drop demo please? Thanks!

**Reply by epezent** [reply]
It’s in implot_demo.cpp

**Reply by djbarker** [reply]
So it is, thanks. I missed it in there.

**Comment by epezent** [comment]
## Colormap Overhaul - (v0.9 - 3/17/2021)

Colormaps have been given a significant overhaul in terms of the public API, features, and internal implementation. These changes were mostly added to support improvements to heatmaps and the addition of 2D histograms (see next post).

### New Features

- Colormaps may now be either *qualitative* (i.e. discrete) or *non-qualitative* (i.e. continuous).  The distinction has no effect on single-item plots (e.g. `PlotLine`), but affects the ways colors are sampled for heatmaps and eventually other plot types that use an entire colormap. When non-qualitative colormaps are added, the provided colors are treated as keys and ImPlot builds a lookup table to perform color interpolation:

![rgb](https://user-images.githubusercontent.com/29577475/111465164-bceebb80-86ef-11eb-9ba6-f1ea6fff4a06.gif)


- 6 new colormaps were added: `ImPlotColormap_Twilight`, `ImPlotColormap_RdBu`, `ImPlotColormap_BrBG`, `ImPlotColormap_PiYG`, `ImPlotColormap_Spectral`, `ImPlotColormap_Greys`
- `ImPlotColormap_Deafult` was removed, and `ImPlotColormap_Deep` is now the deafult.

![colormaps](https://user-images.githubusercontent.com/29577475/111464421-c0357780-86ee-11eb-9191-58e4663ab98b.PNG)

![cmaps](https://user-images.githubusercontent.com/29577475/111464477-cfb4c080-86ee-11eb-8050-4791f3c79896.gif)

- New colormap-centric widgets have been added or improved: `ColormapButton`,  `ColormapSlider`, `ColormapScale`, `ColormapIcon`:

![colormap_utils](https://user-images.githubusercontent.com/29577475/111464927-6f724e80-86ef-11eb-9325-c503f3ec003b.PNG)


### API Changes

- `SetColormap` and `PushColormap(ImVec4*)` have been removed in favor of `AddColormap`. These functions return an `ImPlotColormap` index which can be used with `PushColormap`. Alternatively, you can also push colormaps by the string name provided. See `implot.h` documentation for more details.

```cpp
ImPlotColormap AddColormap(const char* name, const ImVec4* cols, int size, bool qual=true);
ImPlotColormap AddColormap(const char* name, const ImU32*  cols, int size, bool qual=true);

```

- `LerpColormap` was changed to `SampleColormap`. 
- `ColormapScale` now requires a label and ImVec2 size. The label will be displayed as an axis label (see above).

### Implementation Details

- Colormaps are now stored as `Imu32` to save space and make color mixing more efficient
- The memory footprint for each colormap is `size*4` bytes for qualitative maps, or `((size-1)*255+1)*4` for non-qualitative maps. The total memory required for the built-in colormaps is approximately 113 kB, allocated when the context is created.
- Lookup tables provide approximately 20-30% performance when sampling colormaps. This is particularly important for high-density heatmaps (see next post).

**Comment by epezent** [comment]
## PlotHistogram, PlotHistogram2D, and Heatmap Performance - (v0.9 - 3/17/2021)

- APIs for plotting 1D and 2D histograms are now available:

```cpp
// Plots a horizontal histogram. #bins can be a positive integer or an ImPlotBin_ method. If #cumulative is true, each bin contains its count plus the counts of all previous bins.
// If #density is true, the PDF is visualized. If both are true, the CDF is visualized. If #range is left unspecified, the min/max of #values will be used as the range.
// If #range is specified, outlier values outside of the range are not binned. However, outliers still count toward normalizing and cumulative counts unless #outliers is false. The largest bin count or density is returned.
template <typename T> IMPLOT_API double PlotHistogram(const char* label_id, const T* values, int count, int bins=ImPlotBin_Sturges, bool cumulative=false, bool density=false, ImPlotRange range=ImPlotRange(), bool outliers=true, double bar_scale=1.0);

// Plots two dimensional, bivariate histogram as a heatmap. #x_bins and #y_bins can be a positive integer or an ImPlotBin. If #density is true, the PDF is visualized.
// If #range is left unspecified, the min/max of #xs an #ys will be used as the ranges. If #range is specified, outlier values outside of range are not binned.
// However, outliers still count toward the normalizing count for density plots unless #outliers is false. The largest bin count or density is returned.
template <typename T> IMPLOT_API double PlotHistogram2D(const char* label_id, const T* xs, const T* ys, int count, int x_bins=ImPlotBin_Sturges, int y_bins=ImPlotBin_Sturges, bool density=false, ImPlotLimits range=ImPlotLimits(), bool outliers=true);
```

- 5 automatic binning methods are provided: 

```cpp
// Enums for different automatic histogram binning methods (k = bin count or w = bin width)
enum ImPlotBin_ {
    ImPlotBin_Sqrt    = -1, // k = sqrt(n)
    ImPlotBin_Sturges = -2, // k = 1 + log2(n)
    ImPlotBin_Rice    = -3, // k = 2 * cbrt(n)
    ImPlotBin_Scott   = -4, // w = 3.49 * sigma / cbrt(n)
};
```
- enable `density` and/or `cumulative` to plot PDFs and CDFs

![hist](https://user-images.githubusercontent.com/29577475/111469260-7c457100-86f4-11eb-8a6b-20d08a834c44.gif)

![hist2d](https://user-images.githubusercontent.com/29577475/111469276-80718e80-86f4-11eb-89b0-43fb56e96c1d.gif)

- The changes to the colormap implementation (see above) and a number of other improvements have made heatmaps significantly more performant than before (expect at least a 50% improvement in FPS for large heatmaps). As a proof of concept, here is a real-time spectrogram application that renders a 430x513 heatmap. My naïve implementation is able to compute the FFTs and plot the data in the same thread consistently above 60 FPS. 

https://user-images.githubusercontent.com/29577475/111469637-ea8a3380-86f4-11eb-9cee-342f4b4206bd.mp4

**Reply by epezent** [reply]
FYI: The spectrogram application is available here: 
https://github.com/epezent/implot_demos/blob/master/demos/spectrogram.cpp

**Comment by epezent** [comment]
## Initial Fit and Auto Fitting - (v0.10 - 3/20/2021)

By default, plot axes limits are now conveniently fit to data plotted on the first rendered frame. The previous behavior, which always initialized axes limits to [0,1] can be reenabled with `ImPlotAxisFlags_NoInitialFit`. If axis limits are explicitly set with `SetNextPlotLimits` etc, then no initial fitting will occur:

```cpp
float data[] = {1,2};

// new behavior - initial fit
if (ImPlot::BeginPlot("Plot1")) {
  ImPlot::PlotLine("data",data,data,2);
  ImPlot::EndPlot();
}

// old behavior - no initial fit
if (ImPlot::BeginPlot("Plot2",0,0,{-1,0},0,ImPlotAxisFlags_NoInitialFit,ImPlotAxisFlags_NoInitialFit)) {
  ImPlot::PlotLine("data",data,data,2);
  ImPlot::EndPlot();
}

// initial fit ignored if explicit limits given
ImPlot::SetNextPlotLimits(0,2,0,2);
if (ImPlot::BeginPlot("Plot3")) {
  ImPlot::PlotLine("data",data,data,2);
  ImPlot::EndPlot();
}
```

![initfit](https://user-images.githubusercontent.com/29577475/111894395-139a1500-89d0-11eb-8ca0-9d318d7dcca8.gif)

It is also worth noting that `ImPlotAxisFlags_AutoFit` is now available. When set, the axis will continuously fit to all plotted data each frame, and zoom/panning will be disabled. This is particularly useful for heatmaps and 2D histograms.

**Comment by epezent** [comment]
## Custom Label Formatting - (v0.10 - 3/25/2021)

Custom label format specifiers can be provided for each axis now using the following API. The format will be used for tick labels, mouse position, and potentially other numeric displays in the future. The default format is `"%g"`.

```cpp
// Set the format for numeric X axis labels (default="%g"). Formated values will be doubles (i.e. don't supply %d, %i, etc.). Not applicable if ImPlotAxisFlags_Time enabled.
void SetNextPlotFormatX(const char* fmt);
// Set the format for numeric Y axis labels (default="%g"). Formated values will be doubles (i.e. don't supply %d, %i, etc.).
void SetNextPlotFormatY(const char* fmt, ImPlotYAxis y_axis=ImPlotYAxis_1);
```

![ticks](https://user-images.githubusercontent.com/29577475/112497029-d83e7400-8d4a-11eb-873c-225e3d4724d6.png)

**Comment by epezent** [comment]
## Aligned Plots - (v0.10 - 7/7/2021)

A log awaited PR from @ozlb has finally been merged. The functions below allow you to align the pixel padding of adjacent plots so that plot regions remain aligned. This is particularly useful for stacked plots that share a common x-axis:

```cpp
// Align axis padding over multiple plots in a single row or column. If this function returns true, EndAlignedPlots() must be called. #group_id must be unique.
bool BeginAlignedPlots(const char* group_id, ImPlotOrientation orientation = ImPlotOrientation_Vertical);
// Only call EndAlignedPlots() if BeginAlignedPlots() returns true!
void EndAlignedPlots();
```

![aligned](https://user-images.githubusercontent.com/29577475/124862278-f03cab00-df69-11eb-8d22-8159273d4872.gif)

**Comment by epezent** [comment]
## Subplots - (v0.10 - 7/7/2021)

A new subplot API is available! Easily create 2D cells of plots using `BeginSubplots` and `EndSubplots`:

```cpp
bool BeginSubplots(const char* title_id,
                   int rows,
                   int cols,
                   const ImVec2& size,
                   ImPlotSubplotFlags flags = ImPlotSubplotFlags_None,
                   float* row_ratios        = NULL,
                   float* col_ratios        = NULL);

void EndSubplots();
```

Once inside of a subplot context, just use `Begin/EndPlot` as your normally would! Plots are automatically added in row major order, or you can switch to column major if you need to:

```cpp
if (BeginSubplots("My Subplot",2,3,ImVec2(800,400)) {
    for (int i = 0; i < 6; ++i) {
        if (BeginPlot(...)) {
            ImPlot::PlotLine(...);
            ...
            EndPlot();
        }
    }
    EndSubplots();
}
```

Subplot cells can be sized by providing row and column ratios, and provide resizable grips for end users. The ratios passed will out updated accordingly if users interact with resize grips, which you can disable entirely if desired. 

![subplots1](https://user-images.githubusercontent.com/29577475/124862866-fda66500-df6a-11eb-9295-8bd47444e727.gif)

One cool feature of subplots is the ability to share items on a common legend. Of course, you can position the legend or set up drag and drop capabilities just as you can with individual plots:

![subplots2](https://user-images.githubusercontent.com/29577475/124863345-c5ebed00-df6b-11eb-9c0f-e9b6f9c9c15a.gif)

Finally, subplot axes can be easily linked, either on a row, column, or axis basis. The padding of adjacent subplots will be kept in alignment, as described in the previous post:

![subplots3](https://user-images.githubusercontent.com/29577475/124863551-1a8f6800-df6c-11eb-9562-7467548d9e3c.gif)

**Comment by epezent** [comment]
## Demo Revamp - (v0.10 - 7/7/2021)

ImPlot keeps growing, and so does our demo! To make things a bit more navigable, we've adopted tab organization for the demo. 

![demo](https://user-images.githubusercontent.com/29577475/124864231-6abafa00-df6d-11eb-965c-a1a2fff41b3b.gif)

Also, code for demo examples are now more consumable in a self-contained function. Functions can be searched for using the example's header string, e.g.:

```cpp
void ShowDemo_StairstepPlots() {
    static float ys1[101], ys2[101];
    for (int i = 0; i < 101; ++i) {
        ys1[i] = 0.5f + 0.4f * sinf(50 * i * 0.01f);
        ys2[i] = 0.5f + 0.2f * sinf(25 * i * 0.01f);
    }
    if (ImPlot::BeginPlot("Stairstep Plot", "x", "f(x)")) {
        ImPlot::PlotStairs("Signal 1", ys1, 101, 0.01f);
        ImPlot::SetNextMarkerStyle(ImPlotMarker_Square, 2.0f);
        ImPlot::PlotStairs("Signal 2", ys2, 101, 0.01f);
        ImPlot::EndPlot();
    }
}
```

**Comment by epezent** [comment]
## Setup API, Axes Features, and Tools - (v0.13 - 10/19/2021)

It's been quiet around here for several months but not due to a lack of progress. Tons of work has been poured into the new Setup API and accompanying features, and today we are rolling out the first batch of changes. 

### Setup API (see #272)

Previously, setting up a plot and its axes was done by passing several args to `BeginPlot` and leveraging a handful of preceding `SetNextXXX` functions. The former was becoming unwieldy with `BeginPlot` requiring additional args for each newly introduced feature, while the latter was always awkward and a bit of a nightmare to manage under the hood. Neither approach offered room to scale the API gracefully. 

Taking a page out of ImGui's `Tables` API, plots and axes can now be configured by calling a host of **optional**  `SetupXXX` functions immediately after `BeginPlot`. With this, the signature of `BeginPlot` has be greatly simplified. Here's a quick example:

```cpp
if (BeginPlot("My Plot",ImVec2(-1,-1),ImPlotFlags_Crosshairs) { // 1) begin a new plot
  SetupAxis(ImAxis_X1, "Time", ImPlotAxisFlags_Time);           // 2) make Setup calls
  SetupAxis(ImAxis_Y1, "My Y-Axis");
  SetupAxisLimits(ImAxis_Y1, 0, 1000);
  SetupAxisFormat(ImAxis_Y1, "$%.0f");
  SetupLegend(ImPlotLocation_North);
  ...
  SetupFinish();                                                // 3) [optional] explicitly finish setup
  PlotLine(...);                                                // 4) plot items
  ...
  EndPlot();                                                    // 5) end the plot
}
```

All of the functionality you previously had with the original signature of `BeginPlot` and `SetNextXXX` is now available through the Setup API:

```cpp
// Begin a new plot. 
bool BeginPlot(const char* title_id, const ImVec2& size = ImVec2(-1,0), ImPlotFlags flags = ImPlotFlags_None);
```
```cpp
// Enables an axis or sets the label and/or flags for an existing axis. Leave #label = NULL for no label.
void SetupAxis(ImAxis axis, const char* label = NULL, ImPlotAxisFlags flags = ImPlotAxisFlags_None);
// Sets an axis range limits. If ImPlotCond_Always is used, the axes limits will be locked.
void SetupAxisLimits(ImAxis axis, double v_min, double v_max, ImPlotCond cond = ImPlotCond_Once);
// Links an axis range limits to external values. Set to NULL for no linkage. The pointer data must remain valid until EndPlot.
void SetupAxisLinks(ImAxis axis, double* link_min, double* link_max);
// Sets the format of numeric axis labels via formater specifier (default="%g"). Formated values will be double (i.e. use %f).
void SetupAxisFormat(ImAxis axis, const char* fmt);
// Sets the format of numeric axis labels via formatter callback. Given #value, write a label into #buff. Optionally pass user data.
void SetupAxisFormat(ImAxis axis, ImPlotFormatter formatter, void* data = NULL);
// Sets an axis' ticks and optionally the labels. To keep the default ticks, set #keep_default=true.
void SetupAxisTicks(ImAxis axis, const double* values, int n_ticks, const char* const labels[] = NULL, bool keep_default = false);
// Sets an axis' ticks and optionally the labels for the next plot. To keep the default ticks, set #keep_default=true.
void SetupAxisTicks(ImAxis axis, double v_min, double v_max, int n_ticks, const char* const labels[] = NULL, bool keep_default = false);

// Sets the label and/or flags for primary X and Y axes (shorthand for two calls to SetupAxis).
void SetupAxes(const char* x_label, const char* y_label, ImPlotAxisFlags x_flags = ImPlotAxisFlags_None, ImPlotAxisFlags y_flags = ImPlotAxisFlags_None);
// Sets the primary X and Y axes range limits. If ImPlotCond_Always is used, the axes limits will be locked (shorthand for two calls to SetupAxisLimits).
void SetupAxesLimits(double x_min, double x_max, double y_min, double y_max, ImPlotCond cond = ImPlotCond_Once);

// Sets up the plot legend.
void SetupLegend(ImPlotLocation location, ImPlotLegendFlags flags = ImPlotLegendFlags_None);
// Set the location of the current plot's mouse position text (default = South|East).
void SetupMouseText(ImPlotLocation location, ImPlotMouseTextFlags flags = ImPlotMouseTextFlags_None);

// Explicitly finalize plot setup. Once you call this, you cannot make anymore Setup calls for the current plot!
// Note that calling this function is OPTIONAL; it will be called by the first subsequent setup-locking API call.
void SetupFinish();
```

A few notes:
- Always call Setup code at the top of your `BeginPlot` conditional statement.
- Setup is locked once you start plotting or explicitly call `SetupFinish`. Do NOT call Setup code after you begin plotting or after you make any non-Setup API calls (e.g. utils like `PlotToPixels` also lock Setup)
- Calling `SetupFinish` is **OPTIONAL**, but probably good practice. If you do not all it yourself, then the first subsequent plotting or utility function will call it for you.
- The primary X and Y axis (`ImAxis_X1` and `ImAxis_Y1`) are always enabled. To enable auxiliary axes, you must call, e.g. `SetupAxis(ImAxis_Y2,...)`. This replaces the existing method of enabling auxiliary axes, i.e. passing `ImPlotFlags_YAxis2/3`.
- Some `SetNextXXX` API has been removed or renamed; check the top of `implot.cpp` for changelogs
- **The original `BeginPlot` signature has been deprecated, and usage will result in compiler warnings. Please begin transitioning your code before its removal v1.0! You can define `IMPLOT_DISABLE_OBSOLETE_FUNCTIONS` to remove it now.**

### New Axes Features

The new Setup API allows us to easily extend and add functionality to axes. Here are a few new features you can start using today, with more expected to come in the near future.

- support for multiple x-axes (up to 3 x-axes, and 3 y-axes)
  - `ImAxis_` enums replace `ImPlotYAxis_`
  - `SetAxis` replaces `SetPlotYAxis`
- place axes on opposite side with `ImPlotAxisFlags_Opposite` or simply by dragging the axis to the other side
- new style colors: `ImPlotCol_AxisBgHovered` and `ImPlotCol_AxisBgActive`
- when multiple x or y axes are enabled, hovering an item legend icon highlights the axes it is plotted on
- axes use `ImGui::ButtonBehavior` under the hood for more robust interaction

![axes2](https://user-images.githubusercontent.com/29577475/137951482-6c708f9e-59a5-4e37-9533-31d59d0d75c9.gif)

- custom tick label formatting via `ImPlotFormatter` callbacks passed to `SetupAxisFormat`
- formatting is applied to mouse position text
- context menus now display axis label name, if available

![fmt](https://user-images.githubusercontent.com/29577475/137953519-d46cef58-68c4-4251-bec4-464d52be105a.gif)

### Improved Tooling

- `Query` has been removed and replaced with a more flexible tool, `DragRect`. Unlike the previous query system, these can be resized and multiple can exist in one plot. An example in the demo shows how you can convert plot selections in to new query rects.

![dragect](https://user-images.githubusercontent.com/29577475/137954112-4cc9d490-f7c4-4a01-9753-55a9b113462a.gif)
![dragect2](https://user-images.githubusercontent.com/29577475/137954135-1c82bcc8-ef4a-42e0-94fa-749eb02bad06.gif)

- `TagX` and `TagY` have been added and allow you to add custom labels to axes. These use either the axis' formatter, or a custom provided string. You can combine these with other tools like `DragLine` to create custom interactivity.

![tags](https://user-images.githubusercontent.com/29577475/137954599-9075c054-8a68-480b-ba7f-fd2b5cb05b36.gif)

- `DragPoint` and `DragLine` no longer take a string ID. Instead they now expect an `int` ID, and as a result no longer display the optional label. Instead, use `TagX/Y` and `Annotation` to add labels too these tools if needed.
- additional options for drag tools via `ImPlotDragToolFlags`

### Return of ImPlotInputMap

- `ImPlotInputMap` has been revamped and brought back into the public API, and two presets are now provided: `Default` and `Reverse`

```cpp
struct ImPlotInputMap {
    ImGuiMouseButton Pan;           // LMB    enables panning when held,
    ImGuiKeyModFlags PanMod;        // none   optional modifier that must be held for panning/fitting
    ImGuiMouseButton Fit;           // LMB    initiates fit when double clicked
    ImGuiMouseButton Select;        // RMB    begins box selection when pressed and confirms selection when released
    ImGuiMouseButton SelectCancel;  // LMB    cancels active box selection when pressed; cannot be same as Select
    ImGuiKeyModFlags SelectMod;     // none   optional modifier that must be held for box selection
    ImGuiKeyModFlags SelectHorzMod; // Alt    expands active box selection horizontally to plot edge when held
    ImGuiKeyModFlags SelectVertMod; // Shift  expands active box selection vertically to plot edge when held
    ImGuiMouseButton Menu;          // RMB    opens context menus (if enabled) when clicked
    ImGuiKeyModFlags OverrideMod;   // Ctrl   when held, all input is ignored; used to enable axis/plots as DND sources
    ImGuiKeyModFlags ZoomMod;       // none   optional modifier that must be held for scroll wheel zooming
    float            ZoomRate;      // 0.1f   zoom rate for scroll (e.g. 0.1f = 10% plot range every scroll click); make negative to invert
    ImPlotInputMap();
};

// Provides access to input mapping structure for permanent modifications to controls for pan, select, etc.
ImPlotInputMap& GetInputMap();

// Default input mapping: pan = LMB drag, box select = RMB drag, fit = LMB double click, context menu = RMB click, zoom = scroll.
void MapInputDefault(ImPlotInputMap* dst = NULL);
// Reverse input mapping: pan = RMB drag, box select = LMB drag, fit = LMB double click, context menu = RMB click, zoom = scroll.
void MapInputReverse(ImPlotInputMap* dst = NULL);
```
